File: ocaml/d2-2.ml
let rec process_lines () = 
  try
    let line = read_line () in
    let a = String.get line 0 in
    let b = String.get line 2 in
    process_lines () + match b with
    | 'X' -> 0 + (
      match a with
      | 'A' -> 3
      | 'B' -> 1
      | 'C' -> 2
      | _ -> 0)
    | 'Y' -> 3 + (
      match a with
      | 'A' -> 1
      | 'B' -> 2
      | 'C' -> 3
      | _ -> 0)
    | 'Z' -> 6 + (
      match a with
      | 'A' -> 2
      | 'B' -> 3
      | 'C' -> 1
      | _ -> 0)
    | _ -> 0
  with
  _ -> 0;;
  
print_int (process_lines ());;

File: rust/d2-2.rs
use std::io::{self, BufRead};

fn main() -> io::Result<()> {
    let lines = io::stdin().lock().lines();
    let mut score = 0;
    for line in lines {
        let line = line?;
        let mut chars = line.chars();
        let a = chars.nth(0);
        let b = chars.nth(1);
        match b.unwrap() {
            'X' => {
                score += 0;
                match a.unwrap() {
                    'A' =>  score += 3,
                    'B' =>  score += 1,
                    'C' =>  score += 2,
                    _ => println!("Invalid character"),
                }
            },
            'Y' => {
                score += 3;
                match a.unwrap() {
                    'A' =>  score += 1,
                    'B' =>  score += 2,
                    'C' =>  score += 3,
                    _ => println!("Invalid character"),
                }
            },
            'Z' => {
                score += 6;
                match a.unwrap() {
                    'A' =>  score += 2,
                    'B' =>  score += 3,
                    'C' =>  score += 1,
                    _ => println!("Invalid character"),
                }
            },
            _ => println!("Invalid character"),
        }
    }
    println!("{}", score);
    Ok(())
}
